{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## META-ANALYSIS: Applying a Random Effects Model with inverse variance weighting "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### summary of the steps undertaken:\n",
    "    1. Import data (previously generated Hedges' G values and Variances (Vg values))\n",
    "    2. Calculate Combined (Mean) Effect size (M*) for each MGI gene symbol\n",
    "    3. Calculate Z-scores and p values for each gene symbol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  es = effect sizes, imported from indexed file (Hedges g)\n",
    "  \n",
    "  W = weight  (inverse of Variance in Hedges g)\n",
    "  $$ \\frac{1}{var}$$\n",
    "  \n",
    "  M* = Combined (mean) effect size\n",
    "  \n",
    "  W* = weighting in random effects model (inverse of study variance + between studies variance)\n",
    "   $$ \\frac{1}{var + Tau^2} $$\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With\n",
    "\n",
    "$$  {Tau^2\\ } = \\frac{Q - df}{C}  $$  \n",
    "\n",
    "Where:\n",
    "\n",
    "$${ Q\\ }= \\sum{es^2*W} - \\frac{(\\sum{es*W})^2}{\\sum{W}} $$\n",
    "\n",
    "\n",
    "$$ C= \\sum{W} -\\frac{\\sum{W^2}}{\\sum{W}} $$ "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and I squared statistic (heterogeneity within the meta-analysis, see Higgins et al 2003 BMJ, 327:557-60)\n",
    "$$  {I^2\\ } = \\frac{Q^2 - df}{Q}  $$  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## set up "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd # Dataframes and File IO\n",
    "import numpy as np # numerical calculations\n",
    "import matplotlib.pyplot as plt # Plotting\n",
    "import scipy.stats as st # statistical package (p_vals)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MoEx_Enrich</th>\n",
       "      <th>MoEx_Hedges_g</th>\n",
       "      <th>MoEx_Var_g</th>\n",
       "      <th>MoEx_SEg</th>\n",
       "      <th>MoGene_Enrich</th>\n",
       "      <th>MoGene_Hedges_g</th>\n",
       "      <th>MoGene_Var_g</th>\n",
       "      <th>MoGene_SEg</th>\n",
       "      <th>RNA_Seq_Enrich</th>\n",
       "      <th>RNA_Seq_Hedges_g</th>\n",
       "      <th>RNA_Seq_Var_g</th>\n",
       "      <th>RNA_Seq_SEg</th>\n",
       "      <th>430V2_Enrich</th>\n",
       "      <th>430V2_Hedges_g</th>\n",
       "      <th>430V2_Var_g</th>\n",
       "      <th>430V2_SEg</th>\n",
       "      <th>430AV2_Enrich</th>\n",
       "      <th>430AV2_Hedges_g</th>\n",
       "      <th>430AV2_Var_g</th>\n",
       "      <th>430AV2_SEg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.703463</td>\n",
       "      <td>1.705831</td>\n",
       "      <td>0.252462</td>\n",
       "      <td>0.502456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015697</td>\n",
       "      <td>0.569601</td>\n",
       "      <td>0.230165</td>\n",
       "      <td>0.479756</td>\n",
       "      <td>-0.065548</td>\n",
       "      <td>-0.308451</td>\n",
       "      <td>0.373230</td>\n",
       "      <td>0.610925</td>\n",
       "      <td>0.071840</td>\n",
       "      <td>0.392799</td>\n",
       "      <td>0.399175</td>\n",
       "      <td>0.631803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>-0.329912</td>\n",
       "      <td>-1.349822</td>\n",
       "      <td>0.233036</td>\n",
       "      <td>0.482738</td>\n",
       "      <td>-0.370831</td>\n",
       "      <td>-0.995738</td>\n",
       "      <td>0.176630</td>\n",
       "      <td>0.420274</td>\n",
       "      <td>-0.003165</td>\n",
       "      <td>-0.015585</td>\n",
       "      <td>0.220630</td>\n",
       "      <td>0.469713</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610008F07Rik</th>\n",
       "      <td>0.284077</td>\n",
       "      <td>0.610090</td>\n",
       "      <td>0.207147</td>\n",
       "      <td>0.455134</td>\n",
       "      <td>-0.136057</td>\n",
       "      <td>-0.463423</td>\n",
       "      <td>0.160449</td>\n",
       "      <td>0.400560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.029420</td>\n",
       "      <td>-0.177580</td>\n",
       "      <td>0.371242</td>\n",
       "      <td>0.609296</td>\n",
       "      <td>-0.034139</td>\n",
       "      <td>-0.240750</td>\n",
       "      <td>0.394796</td>\n",
       "      <td>0.628328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B14Rik</th>\n",
       "      <td>0.187861</td>\n",
       "      <td>1.120296</td>\n",
       "      <td>0.222912</td>\n",
       "      <td>0.472136</td>\n",
       "      <td>-0.252100</td>\n",
       "      <td>-1.107915</td>\n",
       "      <td>0.181547</td>\n",
       "      <td>0.426083</td>\n",
       "      <td>-0.007089</td>\n",
       "      <td>-0.196339</td>\n",
       "      <td>0.221757</td>\n",
       "      <td>0.470910</td>\n",
       "      <td>-0.326035</td>\n",
       "      <td>-1.337117</td>\n",
       "      <td>0.426128</td>\n",
       "      <td>0.652785</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.008482</td>\n",
       "      <td>0.020895</td>\n",
       "      <td>0.200508</td>\n",
       "      <td>0.447781</td>\n",
       "      <td>-0.141382</td>\n",
       "      <td>-0.267000</td>\n",
       "      <td>0.157460</td>\n",
       "      <td>0.396812</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.744853</td>\n",
       "      <td>0.236941</td>\n",
       "      <td>0.486766</td>\n",
       "      <td>0.392765</td>\n",
       "      <td>1.473526</td>\n",
       "      <td>0.438109</td>\n",
       "      <td>0.661898</td>\n",
       "      <td>0.951853</td>\n",
       "      <td>5.058482</td>\n",
       "      <td>1.555264</td>\n",
       "      <td>1.247102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               MoEx_Enrich  MoEx_Hedges_g  MoEx_Var_g  MoEx_SEg  \\\n",
       "0610005C13Rik     0.703463       1.705831    0.252462  0.502456   \n",
       "0610007P14Rik    -0.329912      -1.349822    0.233036  0.482738   \n",
       "0610008F07Rik     0.284077       0.610090    0.207147  0.455134   \n",
       "0610009B14Rik     0.187861       1.120296    0.222912  0.472136   \n",
       "0610009B22Rik     0.008482       0.020895    0.200508  0.447781   \n",
       "\n",
       "               MoGene_Enrich  MoGene_Hedges_g  MoGene_Var_g  MoGene_SEg  \\\n",
       "0610005C13Rik            NaN              NaN           NaN         NaN   \n",
       "0610007P14Rik      -0.370831        -0.995738      0.176630    0.420274   \n",
       "0610008F07Rik      -0.136057        -0.463423      0.160449    0.400560   \n",
       "0610009B14Rik      -0.252100        -1.107915      0.181547    0.426083   \n",
       "0610009B22Rik      -0.141382        -0.267000      0.157460    0.396812   \n",
       "\n",
       "               RNA_Seq_Enrich  RNA_Seq_Hedges_g  RNA_Seq_Var_g  RNA_Seq_SEg  \\\n",
       "0610005C13Rik        0.015697          0.569601       0.230165     0.479756   \n",
       "0610007P14Rik       -0.003165         -0.015585       0.220630     0.469713   \n",
       "0610008F07Rik             NaN               NaN            NaN          NaN   \n",
       "0610009B14Rik       -0.007089         -0.196339       0.221757     0.470910   \n",
       "0610009B22Rik        0.260870          0.744853       0.236941     0.486766   \n",
       "\n",
       "               430V2_Enrich  430V2_Hedges_g  430V2_Var_g  430V2_SEg  \\\n",
       "0610005C13Rik     -0.065548       -0.308451     0.373230   0.610925   \n",
       "0610007P14Rik           NaN             NaN          NaN        NaN   \n",
       "0610008F07Rik     -0.029420       -0.177580     0.371242   0.609296   \n",
       "0610009B14Rik     -0.326035       -1.337117     0.426128   0.652785   \n",
       "0610009B22Rik      0.392765        1.473526     0.438109   0.661898   \n",
       "\n",
       "               430AV2_Enrich  430AV2_Hedges_g  430AV2_Var_g  430AV2_SEg  \n",
       "0610005C13Rik       0.071840         0.392799      0.399175    0.631803  \n",
       "0610007P14Rik            NaN              NaN           NaN         NaN  \n",
       "0610008F07Rik      -0.034139        -0.240750      0.394796    0.628328  \n",
       "0610009B14Rik            NaN              NaN           NaN         NaN  \n",
       "0610009B22Rik       0.951853         5.058482      1.555264    1.247102  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv('input_files/Symbol_forMeta.csv', index_col=0)# mport indexd values from studies in meta-analysis \n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df['W1'] = 1/df['MoEx_Var_g'] # weight as 1/variance\n",
    "df['W2'] = 1/df['MoGene_Var_g']\n",
    "df['W3'] = 1/df['430V2_Var_g']\n",
    "df['W4'] = 1/df['430AV2_Var_g']\n",
    "df['W5'] = 1/df['RNA_Seq_Var_g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['W1^2'] = np.square(df['W1']) # squared weights\n",
    "df['W2^2'] = np.square(df['W2'])\n",
    "df['W3^2'] = np.square(df['W3'])\n",
    "df['W4^2'] = np.square(df['W4'])\n",
    "df['W5^2'] = np.square(df['W5'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['esSqW1'] = np.square(df['MoEx_Hedges_g'])*df['W1']\n",
    "df['esSqW2'] = np.square(df['MoGene_Hedges_g'])*df['W2']\n",
    "df['esSqW3'] = np.square(df['430V2_Hedges_g'])*df['W3']\n",
    "df['esSqW4'] = np.square(df['430AV2_Hedges_g'])*df['W4']\n",
    "df['esSqW5'] = np.square(df['RNA_Seq_Hedges_g'])*df['W5']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['esW1'] = df['MoEx_Hedges_g']*df['W1']\n",
    "df['esW2'] = df['MoGene_Hedges_g']*df['W2']\n",
    "df['esW3'] = df['430V2_Hedges_g']*df['W3']\n",
    "df['esW4'] =  df['430AV2_Hedges_g']*df['W4']\n",
    "df['esW5'] = df['RNA_Seq_Hedges_g']*df['W5']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# for each row count degrees of freedom as 'Hedges_g' colums with data in, -1\n",
    "\n",
    "df['df'] = df.iloc[:,[1,5,9,13,17]].count(axis=1)-1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Qa = df.filter(regex='esSq').sum(axis=1)\n",
    "Qb = np.square(df.filter(regex='esW').sum(axis=1))\n",
    "Qc = df.filter(regex='^W\\d$').sum(axis=1)  \n",
    "Cd =  df.filter(regex='^W\\d\\^\\d$').sum(axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['Q'] = Qa -(Qb/Qc)\n",
    "df['C'] = Qc -(Cd/Qc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MoEx_Enrich</th>\n",
       "      <th>MoEx_Hedges_g</th>\n",
       "      <th>MoEx_Var_g</th>\n",
       "      <th>MoEx_SEg</th>\n",
       "      <th>MoGene_Enrich</th>\n",
       "      <th>MoGene_Hedges_g</th>\n",
       "      <th>MoGene_Var_g</th>\n",
       "      <th>MoGene_SEg</th>\n",
       "      <th>RNA_Seq_Enrich</th>\n",
       "      <th>RNA_Seq_Hedges_g</th>\n",
       "      <th>...</th>\n",
       "      <th>esSqW5</th>\n",
       "      <th>esW1</th>\n",
       "      <th>esW2</th>\n",
       "      <th>esW3</th>\n",
       "      <th>esW4</th>\n",
       "      <th>esW5</th>\n",
       "      <th>df</th>\n",
       "      <th>Q</th>\n",
       "      <th>C</th>\n",
       "      <th>Tau^2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.703463</td>\n",
       "      <td>1.705831</td>\n",
       "      <td>0.252462</td>\n",
       "      <td>0.502456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015697</td>\n",
       "      <td>0.569601</td>\n",
       "      <td>...</td>\n",
       "      <td>1.409619</td>\n",
       "      <td>6.756779</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.826437</td>\n",
       "      <td>0.984028</td>\n",
       "      <td>2.474747</td>\n",
       "      <td>3</td>\n",
       "      <td>7.042189</td>\n",
       "      <td>9.930513</td>\n",
       "      <td>0.407047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>-0.329912</td>\n",
       "      <td>-1.349822</td>\n",
       "      <td>0.233036</td>\n",
       "      <td>0.482738</td>\n",
       "      <td>-0.370831</td>\n",
       "      <td>-0.995738</td>\n",
       "      <td>0.176630</td>\n",
       "      <td>0.420274</td>\n",
       "      <td>-0.003165</td>\n",
       "      <td>-0.015585</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001101</td>\n",
       "      <td>-5.792322</td>\n",
       "      <td>-5.637409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.070641</td>\n",
       "      <td>2</td>\n",
       "      <td>4.302481</td>\n",
       "      <td>9.582901</td>\n",
       "      <td>0.240270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610008F07Rik</th>\n",
       "      <td>0.284077</td>\n",
       "      <td>0.610090</td>\n",
       "      <td>0.207147</td>\n",
       "      <td>0.455134</td>\n",
       "      <td>-0.136057</td>\n",
       "      <td>-0.463423</td>\n",
       "      <td>0.160449</td>\n",
       "      <td>0.400560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.945202</td>\n",
       "      <td>-2.888300</td>\n",
       "      <td>-0.478341</td>\n",
       "      <td>-0.609810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3.301801</td>\n",
       "      <td>11.631238</td>\n",
       "      <td>0.025947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B14Rik</th>\n",
       "      <td>0.187861</td>\n",
       "      <td>1.120296</td>\n",
       "      <td>0.222912</td>\n",
       "      <td>0.472136</td>\n",
       "      <td>-0.252100</td>\n",
       "      <td>-1.107915</td>\n",
       "      <td>0.181547</td>\n",
       "      <td>0.426083</td>\n",
       "      <td>-0.007089</td>\n",
       "      <td>-0.196339</td>\n",
       "      <td>...</td>\n",
       "      <td>0.173834</td>\n",
       "      <td>5.025728</td>\n",
       "      <td>-6.102643</td>\n",
       "      <td>-3.137831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.885379</td>\n",
       "      <td>3</td>\n",
       "      <td>15.217340</td>\n",
       "      <td>12.321941</td>\n",
       "      <td>0.991511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.008482</td>\n",
       "      <td>0.020895</td>\n",
       "      <td>0.200508</td>\n",
       "      <td>0.447781</td>\n",
       "      <td>-0.141382</td>\n",
       "      <td>-0.267000</td>\n",
       "      <td>0.157460</td>\n",
       "      <td>0.396812</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.744853</td>\n",
       "      <td>...</td>\n",
       "      <td>2.341539</td>\n",
       "      <td>0.104210</td>\n",
       "      <td>-1.695672</td>\n",
       "      <td>3.363378</td>\n",
       "      <td>3.252492</td>\n",
       "      <td>3.143626</td>\n",
       "      <td>4</td>\n",
       "      <td>20.595754</td>\n",
       "      <td>13.688561</td>\n",
       "      <td>1.212381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               MoEx_Enrich  MoEx_Hedges_g  MoEx_Var_g  MoEx_SEg  \\\n",
       "0610005C13Rik     0.703463       1.705831    0.252462  0.502456   \n",
       "0610007P14Rik    -0.329912      -1.349822    0.233036  0.482738   \n",
       "0610008F07Rik     0.284077       0.610090    0.207147  0.455134   \n",
       "0610009B14Rik     0.187861       1.120296    0.222912  0.472136   \n",
       "0610009B22Rik     0.008482       0.020895    0.200508  0.447781   \n",
       "\n",
       "               MoGene_Enrich  MoGene_Hedges_g  MoGene_Var_g  MoGene_SEg  \\\n",
       "0610005C13Rik            NaN              NaN           NaN         NaN   \n",
       "0610007P14Rik      -0.370831        -0.995738      0.176630    0.420274   \n",
       "0610008F07Rik      -0.136057        -0.463423      0.160449    0.400560   \n",
       "0610009B14Rik      -0.252100        -1.107915      0.181547    0.426083   \n",
       "0610009B22Rik      -0.141382        -0.267000      0.157460    0.396812   \n",
       "\n",
       "               RNA_Seq_Enrich  RNA_Seq_Hedges_g    ...       esSqW5      esW1  \\\n",
       "0610005C13Rik        0.015697          0.569601    ...     1.409619  6.756779   \n",
       "0610007P14Rik       -0.003165         -0.015585    ...     0.001101 -5.792322   \n",
       "0610008F07Rik             NaN               NaN    ...          NaN  2.945202   \n",
       "0610009B14Rik       -0.007089         -0.196339    ...     0.173834  5.025728   \n",
       "0610009B22Rik        0.260870          0.744853    ...     2.341539  0.104210   \n",
       "\n",
       "                   esW2      esW3      esW4      esW5  df          Q  \\\n",
       "0610005C13Rik       NaN -0.826437  0.984028  2.474747   3   7.042189   \n",
       "0610007P14Rik -5.637409       NaN       NaN -0.070641   2   4.302481   \n",
       "0610008F07Rik -2.888300 -0.478341 -0.609810       NaN   3   3.301801   \n",
       "0610009B14Rik -6.102643 -3.137831       NaN -0.885379   3  15.217340   \n",
       "0610009B22Rik -1.695672  3.363378  3.252492  3.143626   4  20.595754   \n",
       "\n",
       "                       C     Tau^2  \n",
       "0610005C13Rik   9.930513  0.407047  \n",
       "0610007P14Rik   9.582901  0.240270  \n",
       "0610008F07Rik  11.631238  0.025947  \n",
       "0610009B14Rik  12.321941  0.991511  \n",
       "0610009B22Rik  13.688561  1.212381  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Tau^2'] = (df['Q']-df['df'])/df['C']\n",
    "df['Tau^2'].where(df['Tau^2']>0, 0,inplace=True) # set Tau-squared to 0 where <0\n",
    "df['Tau^2'].where(df['Tau^2']<100000, 0,inplace=True) # set Tau-squared to 0 where err or inf\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df['I^2'] = (np.square(df['Q'])-df['df'])/df['Q']\n",
    "df['I^2'].where(df['I^2']>0, 0,inplace=True)   # set I-squared to 0 where <0\n",
    "df['I^2'].where(df['I^2']<100000, 0,inplace=True)  # set I-squared to 0 where err or inf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0xa7d38d0>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZAAAAEACAYAAACd2SCPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAHjJJREFUeJzt3X+UXHV9//Hna5Nswi8XQ5FAAogQiFJqRAmpyGGtR0iC\nuj1iz5eo3wi1JVWiVGoltVY2p1aFVr8hRQRqQPCrhiMqBhsRFNevWI1BiWDYQEB+JCHE8JsE8mvz\n/v5xZ2Gc7OzO/rj3zr37epyzZ2bufGbmfdlZXvn8uPcqIjAzMxuslrwLMDOzYnKAmJnZkDhAzMxs\nSBwgZmY2JA4QMzMbEgeImZkNSeoBImmWpLWS7pd0UZ02SyStk7Ra0vSq7W2SviWpW9IaSSenXa+Z\nmTUm1QCR1AJcDpwBHA/MlTStps1s4OiImArMB66sevoyYEVEvBZ4PdCdZr1mZta4tHsgM4B1EfFI\nROwClgEdNW06gOsBImIl0CbpEEmvAE6NiGsrz+2OiOdSrtfMzBqUdoBMBtZXPd5Q2dZfm42VbUcB\nT0i6VtJvJF0taZ9UqzUzs4Y18yT6WOBE4EsRcSLwArAw35LMzKzX2JTffyNwRNXjKZVttW0Or9Nm\nfUTcWbl/I1BvEt4n9DIzG6SI0HBen3YPZBVwjKQjJbUCZwPLa9osB+YBSJoJPBMRmyNiM7Be0rGV\ndm8D7q33QRFRyp+LL7449xrS+nn00QAupqsr/1r8+/P+jbb9Gwmp9kAiokfSAuBWkrBaGhHdkuYn\nT8fVEbFC0hxJDwDbgHOr3uKjwNcljQN+X/OcFdyjjya3Dz8Mp52WaylmNgRpD2EREbcAx9Vsu6rm\n8YI6r/0tcFJ61VmennwyuX3qqXzrMLOhaeZJdAPa29vzLiE1W7cCtPNciRdnl/n3B96/0c4B0uTK\n/AV2gBSf9290c4BYbp5/Hvbdl1IHiFmZOUAsN1u3wmGHJUFiZsXjALHcbN0Kkye7B2JWVA4Qy83z\nzyc9EAeIWTE5QCw37oGYFZsDxHKzdStMmtS7GsvMisYBYrnZtg0mToQdO/KuxMyGwgFiudmxAw48\nELZvz7sSMxsKB4jlZscOaGtzgJgVlQPEclMdICN0clAzy5ADxHKzfXtyJHpLC+zenXc1ZjZYDhDL\nzY4dMH48TJjgYSyzInKAWG4cIGbF5gCx3OzYkYSHA8SsmBwglhv3QMyKzQFiuXGAmBWbA8Ry0dOT\n/Iwd6wAxKyoHiOWit/chOUDMisoBYrnoDRBwgJgVlQPEcuEAMSs+B4jlwgFiVnwOEMuFA8Ss+Bwg\nlgsHiFnxOUAsF7t3w7hxyX0HiFkxOUAsF7t2JceAgAPErKgcIJaLXbvcAzErutQDRNIsSWsl3S/p\nojptlkhaJ2m1pDdUbX9Y0m8l3SXpV2nXatmpDpDx4x0gZkU0Ns03l9QCXA68DXgMWCXpexGxtqrN\nbODoiJgq6WTgy8DMytN7gPaIeDrNOi171XMg48fDM8/kW4+ZDV7aPZAZwLqIeCQidgHLgI6aNh3A\n9QARsRJok3RI5TllUKPloHoOZPz4ZFWWmRVL2v9zngysr3q8obKtvzYbq9oEcJukVZL+NrUqLXO1\nQ1gOELPiSXUIawScEhGbJB1MEiTdEXFH3kXZ8DlAzIov7QDZCBxR9XhKZVttm8P7ahMRmyq3WyR9\nl2RIrM8A6ezsfOl+e3s77e3tw6vcUlU9B9La6gAxS1tXVxddXV0j+p5pB8gq4BhJRwKbgLOBuTVt\nlgPnAzdImgk8ExGbJe0LtETEVkn7AacDi+p9UHWAWPOrnQPZuTPfeszKrvYf1osW1f3facNSDZCI\n6JG0ALiVZL5laUR0S5qfPB1XR8QKSXMkPQBsA86tvPwQ4LuSolLn1yPi1jTrtex4CMus+FKfA4mI\nW4DjarZdVfN4QR+vewiYnm51lhcHiFnxeYms5aL2OBAHiFnxOEAsFz4OxKz4HCCWCw9hmRWfA8Ry\nUR0gra1ehWVWRA4Qy4XnQMyKzwFiufAciFnxOUAsF54DMSs+B4jlwgFiVnwOEMuF50DMis8BYrmo\nngNpbU0e79mTb01mNjgOEMtF9RCWlNzftSvfmsxscBwglovqAAEPY5kVkQPEclE9BwIOELMicoBY\nLqrnQMABYlZEDhDLhYewzIrPAWK5cICYFZ8DxHLhORCz4nOAWC5q50B8Rl6z4nGAWC48hGVWfA4Q\ny4UDxKz4HCCWC8+BmBWfA8Ry4eNAzIrPAWK58BCWWfE5QCwXtQHS2uoAMSsaB4jloq85EC/jNSsW\nB4jlwnMgZsXnALFceA7ErPgcIJYLB4hZ8TlALBc+DsSs+FIPEEmzJK2VdL+ki+q0WSJpnaTVkqbX\nPNci6TeSlqddq2XHcyBmxZdqgEhqAS4HzgCOB+ZKmlbTZjZwdERMBeYDV9a8zQXAvWnWadnzMl6z\n4ku7BzIDWBcRj0TELmAZ0FHTpgO4HiAiVgJtkg4BkDQFmAN8JeU6LWN9zYF4Ga9ZsaQdIJOB9VWP\nN1S29ddmY1Wb/wP8IxBpFWjZ27MnuR0z5uVtHsIyK56xAzfJh6Qzgc0RsVpSO6D+2nd2dr50v729\nnfb29jTLs2Gonf8AB4hZ2rq6uujq6hrR90w7QDYCR1Q9nlLZVtvm8D7avAd4l6Q5wD7AAZKuj4h5\nfX1QdYBYc6sdvgIHiFnaav9hvWjRomG/Z9pDWKuAYyQdKakVOBuoXU21HJgHIGkm8ExEbI6IT0bE\nERHxmsrrbq8XHlYsDhCzcki1BxIRPZIWALeShNXSiOiWND95Oq6OiBWS5kh6ANgGnJtmTZa/2mNA\nwAFiVkSpz4FExC3AcTXbrqp5vGCA9/gp8NORr87y0NcciK+JblY8PhLdMuchLLNycIBY5hwgZuXg\nALHMeQ7ErBwcIJY5HwdiVg4OEMuch7DMysEBYplzgJiVgwPEMtfXHIiX8ZoVjwPEMuc5ELNycIBY\n5vobwgqfd9msMBwglrm+AmTMGGhpSYa3zKwYHCCWub7mQMDDWGZF4wCxzPU1BwIOELOicYBY5voa\nwgJfF92saBwglrl6AeLropsViwPEMuc5ELNycIBY5jwHYlYODQWIpO9IOlOSA8eGrb8hLAeIWXE0\nGghXAO8F1kn6vKTjBnqBWT0OELNyaChAIuJHEfE+4ETgYeBHkv5H0rmS+vhfgVl9ngMxK4eGh6Qk\nHQScA/wNcBdwGUmg3JZKZVZa9eZAvIzXrFj6+DPem6TvAscBXwPeGRGbKk/dIOnOtIqzcqo3hDVh\nggPErEgaChDgvyJiRfUGSeMjYkdEvCmFuqzE6gXIvvvCCy9kX4+ZDU2jQ1if6WPbL0ayEBs96s2B\nOEDMiqXfHoikScBkYB9JbwBUeeoVwL4p12YlVW8OxAFiViwDDWGdQTJxPgX4YtX254FPplSTlVy9\nIaz99nOAmBVJvwESEdcB10k6KyK+nVFNVnL9zYFs25Z9PWY2NAMNYb0/Iv4v8GpJF9Y+HxFf7ONl\nZv3avbv+ENbTT2dfj5kNzUBDWPtVbvdPuxAbPbwKy6wcBhrCuqpyu2ioHyBpFrCYZMXX0oi4pI82\nS4DZwDbgnIhYLWk88P+A1kqdNw6nDmsengMxK4dGT6Z4qaRXSBon6ceStkh6fwOvawEuJ5mMPx6Y\nK2laTZvZwNERMRWYD1wJEBE7gLdGxBuA6cBsSTMGs3PWnPpbxus5ELPiaPQ4kNMj4jngHSTnwjoG\n+McGXjcDWBcRj0TELmAZ0FHTpgO4HiAiVgJtkg6pPO799+h4kl5INFivNTEv4zUrh0YDpPfP/Uzg\nWxHxbIOvmwysr3q8obKtvzYbe9tIapF0F/A4cFtErGrwc62JeQjLrBwaPZXJ9yWtBV4EPiTpYGB7\nemUlImIP8AZJrwBukvS6iLi3r7adnZ0v3W9vb6e9vT3t8myIPIlulr2uri66urpG9D0bCpCIWCjp\nUuDZiOiRtI29h6L6shE4ourxlMq22jaH99cmIp6T9BNgFjBggFhz628Zr+dAzNJR+w/rRYuGvyZp\nMFcYnAb8L0nzgPcApzfwmlXAMZKOlNQKnA0sr2mzHJgHIGkm8ExEbJb0J5LaKtv3Ad4OrB1Evdak\n3AMxK4dGT+f+NeBoYDXQU9kcVCa/66n0VhYAt/LyMt5uSfOTp+PqiFghaY6kB0iW8Z5befmhJEfB\nt1Ree0PtGYGtmDwHYlYOjc6BvAl4XUQMehVURNxCci2R6m1X1Txe0Mfr7iG5YJWVjJfxmpVDo0NY\nvwMmpVmIjR79LePdvh16evZ+zsyaT6M9kD8B7pX0K+Cla8ZFxLtSqcpKrd4QVksL7L8/PP88HHhg\n9nWZ2eA0GiCdaRZho0u9ISxIguOZZxwgZkXQ6DLen0o6EpgaET+StC8wJt3SrKzqDWHBywFiZs2v\n0XNh/S1wI9A7+T0ZuCmtoqzc6g1hgQPErEganUQ/HzgFeA4gItYBr0qrKCu3gQLk2UZPlGNmuWo0\nQHZExM7eB5J8YkMbsv7mQNra3AMxK4pGA+Snkj4J7CPp7cC3gJvTK8vKzHMgZuXQaIAsBLYA95Bc\ns2MF8Km0irJy8xyIWTk0ugprj6SbgJsiYkvKNVnJDbSMd8OGbOsxs6HptweiRKekJ4D7gPsqVyP8\ndDblWRl5CMusHAYawvoYyeqrkyJiYkRMBE4GTpH0sdSrs9LZsyf5GVPnKKKDDoIt7uOaFcJAAfK/\ngbkR8VDvhoj4PfB+KqdgNxuM3vkPqe/nDz0UNm3KtiYzG5qBAmRcRDxRu7EyD1JnFNusvv7mPwAm\nTXKAmBXFQAGyc4jPmfWpv/kPSALkD3/wGXnNimCgVVivl/RcH9sFTEihHiu5/pbwArS2JhPpW7Yk\nYWJmzavfAIkInzDRRtRAQ1gAr3kNPPigA8Ss2Q3mmuhmwzbQEBbAscfCunXZ1GNmQ+cAsUwNNIQF\nSYCsXZtNPWY2dA4Qy1QjAfLGN8KqVdnUY2ZD5wCxTO3ePfAQ1syZSYDs2NF/OzPLlwPEMtVID2Ti\nRHjTm+Bmn+/ZrKk5QCxTjQQIwAc/CIsXQ/iqM2ZNywFimWpkGS/A2WfDzp2wZEn6NZnZ0DR0Onez\nkdLIMl5ITrZ4ww3w5jcnq7Jmz06/NjMbHPdALFONDmEBHHUUfPvb8IEPwMqV6dZlZoPnALFMDSZA\nIOmBXHMNdHTAmjXp1WVmg+cAsUw1soy31jveAV/4AsyaBQ8/nEpZZjYEqQeIpFmS1kq6X9JFddos\nkbRO0mpJ0yvbpki6XdIaSfdI+mjatVr6BtsD6fW+98EnPgFvfzts3jzydZnZ4KUaIJJagMuBM4Dj\ngbmSptW0mQ0cHRFTgfnAlZWndgMXRsTxwJ8D59e+1opnqAEC8JGPJEEyaxY8++zI1mVmg5d2D2QG\nsC4iHomIXcAyoKOmTQdwPUBErATaJB0SEY9HxOrK9q1ANzA55XotZY0u463n4ovhLW+Bd74TXnhh\n5Ooys8FLO0AmA+urHm9g7xCobbOxto2kVwPTAa/FKbhGl/HWI8Fll8HhhycT608/PXK1mdngNP1x\nIJL2B24ELqj0RPrU2dn50v329nba29tTr80GbzhDWL1aWuCrX03mRF7/eujshL/6KzjggJGo0Kyc\nurq66OrqGtH3VKR4rghJM4HOiJhVebwQiIi4pKrNlcBPIuKGyuO1wGkRsVnSWOD7wA8i4rJ+PifS\n3A8bOVdcAffcA1/+8si8389+Bp//fHL71rfCeefBmWeOzHublZkkIkLDeY+0h7BWAcdIOlJSK3A2\nsLymzXJgHrwUOM9ERO86m2uAe/sLDyuWnTth/PiRe79TT4X//m9Yvx7e/W648MIkRPbsGbnPMLO+\npRogEdEDLABuBdYAyyKiW9J8SedV2qwAHpL0AHAV8CEASacA7wP+QtJdkn4jaVaa9Vr6du5Mrns+\n0trakiPWf/1r6O5OJtvNLF2pDmFlxUNYxfGZz8D27cltWh5/HKZPT04Hf9JJ6X2OWZEVYQjL7I+k\n1QOpNmlSElALF6b7OWajnQPEMpVFgEAynPXoo3D77el/ltlo5QCxTGUVIOPGwT//M1x6afqfZTZa\nOUAsU1kFCCQXpVq9GtauzebzzEYbB4hlKssAmTAhWdLrqxqapcMBYpnKMkAAPvQh+OY3fcoTszQ4\nQCxTWQfIoYcmR6YvXZrdZ5qNFg4Qy1TWAQJwwQVw+eXQ05Pt55qVnQPEMrVjR/YBctJJSU/k5puz\n/VyzsnOAWKby6IEAfPSjnkw3G2kOEMtUXgFy1lnJct67787+s83KygFimRrps/E2qrU1WZH1n/+Z\n/WeblZUDxDKVVw8EYP58uPFGePLJfD7frGwcIJapPAPkVa9KLoN79dX5fL5Z2ThALFN5BgjAP/xD\nck31bdvyq8GsLBwglqm8A+SEE5KrGI7UJXXNRjMHiGUq7wAB+PSn4T/+w70Qs+FygFimmiFATjgB\n2tth8eJ86zArOl/S1jL1ylfC73+f3ObpwQfh5JOT66cffHC+tZjlwZe0tcJphh4IwNFHw3vfC//6\nr3lXYlZc7oFYpsaNgxdeSG7ztmULvPa1sHJlEihmo4l7IFYoe/bA7t0wdmzelSQOPhj+/u+TS9+a\n2eC5B2KZ2b4d2tqSM/I2i23b4Nhj4aabkrP2mo0W7oFYoTTL/Ee1/faDzk74xCfA/wYxGxwHiGWm\nGQME4Nxz4fHH4ZZb8q7ErFgcIJaZF1+EffbJu4q9jR2b9EI++9m8KzErFgeIZWb7dpgwIe8q+nbW\nWbBxI/ziF3lXYlYcDhDLTLP2QCDphVx4Ifz7v+ddiVlxpB4gkmZJWivpfkkX1WmzRNI6SaslvaFq\n+1JJmyX5OnIl0MwBAslcyB13wH335V2JWTGkGiCSWoDLgTOA44G5kqbVtJkNHB0RU4H5QPV5Uq+t\nvNZKoJmHsCBZkfXhD8MXvpB3JWbFkHYPZAawLiIeiYhdwDKgo6ZNB3A9QESsBNokHVJ5fAfwdMo1\nWkaavQcCcP75yVULH38870rMml/aATIZWF/1eENlW39tNvbRxkqgCAFy8MEwdy4sWZJ3JWbNz5Po\nlplmH8Lq9fGPJ5e9Xb9+4LZmo1naZyXaCBxR9XhKZVttm8MHaDOgzs7Ol+63t7fT3t4+2LewlBWh\nBwJw1FHwkY/ABRfAd76TdzVmI6Orq4uurq4Rfc9Uz4UlaQxwH/A2YBPwK2BuRHRXtZkDnB8RZ0qa\nCSyOiJlVz78auDkiTujnc3wurAL40pdgzRq44oq8KxnY9u3wZ3+WLOvtqJ21MyuBpj8XVkT0AAuA\nW4E1wLKI6JY0X9J5lTYrgIckPQBcBXy49/WSvgH8D3CspEclnZtmvZauogxhQVLnV76SrMp62ss4\nzPrks/FaZj7zmWQY69/+Le9KGrdgQXL9kmuuybsSs5HV9D0Qs2pFmQOp9rnPwe23ww9/mHclZs3H\nAWKZKdIQVq8DDkjmbC64ILkYlpm9zAFimSliDwRg9mw49FD46lfzrsSsuThALDNFDRAJLrkkOeX7\nCy/kXY1Z83CAWGaKOITVa8YM+PM/h8WL867ErHk4QCwzRe2B9Prc5+CLX4TNm/OuxKw5OEAsM0UP\nkGOOgXnzkqEsM3OAWIZefLG4Q1i9PvUp+Pa3kyPqzUY7B4hl5vnnk2WxRTZxIixaBH/9117Wa+YA\nscxs3Vr8AAH4u7+DAw9Mjqw3G80cIJaZMvRAIFnWe+21sHSpz9Zro1vap3M3e0lZAgTgsMPgpptg\n1iyYPBlOPjnvisyy5x6IZaKnJzkOZN99865k5LzxjUlPpKMDursHbm9WNg4Qy8S2bUl4tJTsG/eO\nd8CllyY9EV/B0EYbD2FZJso0fFVr3jx44gk4/XS44w446KC8KzLLRsn+PWjNqiwrsOq58MJkKGvO\nnGRfzUYDB4hl4rnnyh0gkJzq5IQT4KyzYOfOvKsxS58DxDLx5JPlH9qR4Mork7meD3wA9uzJuyKz\ndDlALBNPPZUcxV12Y8fCN78JmzYlF6HylZatzBwglonR0APpNWECfO978POfw7/8S97VmKXHq7As\nE6OlB9KrrS25jnp7exIon/pU3hWZjTz3QCwToy1AAA4+GH78Y/j61+HjH/eciJWPA8Qy8cQTo2cI\nq9qkSclQ1sqVMHducjS+WVk4QCwTjz2WnD9qNJo4EW67DcaMgTe/GR58MO+KzEaGA8QysWEDTJmS\ndxX5mTAhGcr64AeTa6tfe62HtKz4FCVYZygpyrAfZRWRHBuxZQvsv3/e1eRv9Wo47zwYPx4++1k4\n9dS8K7LRSBIRoeG8h3sglrqnnkr+Z+nwSEyfDr/4BZxzTvLzlrfANdfAs8/mXZnZ4DhALHXd3TBt\nWt5VNJcxY5LhrPvuS86j9f3vw5FHwnvfm8yX9PTkXaHZwFIPEEmzJK2VdL+ki+q0WSJpnaTVkqYP\n5rXW/H73O/jTP827iuY0diy8+93JlQ0ffDCZZF+4EKZOhcWLk3OImTWrVANEUgtwOXAGcDwwV9K0\nmjazgaMjYiowH7iy0deOBl1dXXmXMGx33pkM2/SlDPvXn8Hs30EHwYIF8Otfwze+Ab/8JRx1FHzs\nY/DQQ+nVOBz+/Y1uafdAZgDrIuKRiNgFLAM6atp0ANcDRMRKoE3SIQ2+tvSK/gXu6YEf/CC5VkZf\nir5/Axnq/s2cCcuWwV13QWsrnHQSnHYafPKTSW9lzRrYsWNkax0K//5Gt7RPZTIZqL5O2waSYBio\nzeQGX2tNbunSZGx/6tS8KymmI46ASy6BT386OSDx5z+H665L5k4efjg5tmby5OTnsMNefnzYYcnx\nJ62tMG5c8tN7v3pb2a4QadlqxnNhDWtZWZnce28ylHHnnS+f1bX2ttFtQ31uOO23bUtWYN1yS3Kq\ncxu6/fZLenHVPbmdO+GRR5KDNDduTG4fewxWrUpun3oKdu16+Wfnzr3vjxmzd6j0FzpjxvxxXQ89\nBD/72d719vX7rt3WSJs83qtaW1v95yzl40AkzQQ6I2JW5fFCICLikqo2VwI/iYgbKo/XAqcBRw30\n2qr38EEgZmaDNNzjQNLugawCjpF0JLAJOBuYW9NmOXA+cEMlcJ6JiM2SnmjgtcDw/yOYmdngpRog\nEdEjaQFwK8mE/dKI6JY0P3k6ro6IFZLmSHoA2Aac299r06zXzMwaV4pTmZiZWfYKtQZD0nsk/U5S\nj6QTa577p8rBiN2STq/afqKkuysHIy7OvuqhkXSxpA2SflP5mVX1XJ/7WjRlO1BU0sOSfivpLkm/\nqmx7paRbJd0n6YeSCjMtK2mppM2S7q7aVnd/iva9rLN/pfm7kzRF0u2S1ki6R9JHK9tH7ncYEYX5\nAY4DpgK3AydWbX8tcBfJkNyrgQd4uXe1Ejipcn8FcEbe+9Hgvl4MXNjH9rr7WqQfkn+8PAAcCYwD\nVgPT8q5rmPv0e+CVNdsuAT5RuX8R8Pm86xzE/rwFmA7cPdD+AK8r2veyzv6V5u8OmARMr9zfH7gP\nmDaSv8NC9UAi4r6IWMfeS307gGURsTsiHgbWATMkTQIOiIhVlXbXA3+ZWcHD19figD73NdOqRkYZ\nDxQVe/fqO4DrKvevo0Dfv4i4A3i6ZnO9/XkXBfte1tk/KMnfXUQ8HhGrK/e3At3AFEbwd1ioAOlH\n7UGHG3n5YMQNVdt7D1IsigWV84N9paqbWW9fi6beAaRFFsBtklZJ+pvKtkMiYjMkf9DAq3KrbmS8\nqs7+lOV7CSX8u5P0apLe1i+p/50c9D42XYBIuq0yZ9H7c0/l9p151zbSBtjXK4DXRMR04HHgC/lW\naw04JSJOBOYA50s6lSRUqpVt1UrZ9qd0f3eS9gduBC6o9ERG7DvZdEeiR8Tbh/CyjcDhVY+nVLbV\n294UBrGv/wXcXLnf1Ps0CBuBI6oeF3U/XhIRmyq3WyTdRNL93yzpkEiObZoE/CHXIoev3v6U4nsZ\nEVuqHhb+707SWJLw+FpEfK+yecR+h03XAxmE6nHK5cDZklolHQUcA/yq0j17VtIMSQLmAd/r472a\nTuUX2+vdwO8q9/vc16zrGwEvHWQqqZXkQNHlOdc0ZJL2rfxLD0n7AacD95Ds0zmVZh+gIN+/KmLv\nv7VzKver96eo38s/2r8S/t1dA9wbEZdVbRu532HeKwUGuargL0nG6F4kOTr9B1XP/RPJqoFu4PSq\n7W8k+UNeB1yW9z4MYl+vB+4mWZ10E8m4Zb/7WrQfYBbJypB1wMK86xnmvhxV+V3dVfm+Laxsnwj8\nqLKftwIH5l3rIPbpG8BjwA7gUZKDfF9Zb3+K9r2ss3+l+bsDTgF6qr6Xv6n8zdX9Tg52H30goZmZ\nDUmRh7DMzCxHDhAzMxsSB4iZmQ2JA8TMzIbEAWJmZkPiADEzsyFxgJiZ2ZA4QMzMbEj+P52PDpXI\ngA9TAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0xa59e780>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df['I^2'].plot(kind='kde')  # have a look at distribution of I-squared values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['W*1'] = 1/(df['MoEx_Var_g']+ df['Tau^2'])  # weighting for each study under REM\n",
    "df['W*2'] = 1/(df['MoGene_Var_g']+df['Tau^2'])\n",
    "df['W*3'] = 1/(df['430V2_Var_g']+df['Tau^2'])\n",
    "df['W*4'] = 1/(df['430AV2_Var_g']+df['Tau^2'])\n",
    "df['W*5'] = 1/(df['RNA_Seq_Var_g']+df['Tau^2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#df['W*2']  # check ouput"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['esW*1'] = df['MoEx_Hedges_g']*df['W*1']  #  weighted effect sizes for each study\n",
    "df['esW*2'] = df['MoGene_Hedges_g']*df['W*2']\n",
    "df['esW*3'] = df['430V2_Hedges_g']*df['W*3']\n",
    "df['esW*4'] = df['430AV2_Hedges_g']*df['W*4']\n",
    "df['esW*5'] = df['RNA_Seq_Hedges_g']*df['W*5']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Then sum these weighted effect sizes and weightings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Ma = df.filter(regex='esW\\*\\d').sum(axis=1)  # careful to not summate esW values \n",
    "Mb = df.filter(regex='^W\\*\\d$').sum(axis=1)\n",
    " \n",
    "\n",
    "df['REM_M*'] = Ma/Mb\n",
    "df['REM_SEM*'] = np.sqrt(1/Mb)\n",
    "df['REM_Z'] = df['REM_M*']/df['REM_SEM*']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df['REM_pVal'] = st.norm.sf(abs(df['REM_Z']))*2 # two-sided p-value from Z scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MoEx_Enrich</th>\n",
       "      <th>MoEx_Hedges_g</th>\n",
       "      <th>MoEx_Var_g</th>\n",
       "      <th>MoEx_SEg</th>\n",
       "      <th>MoGene_Enrich</th>\n",
       "      <th>MoGene_Hedges_g</th>\n",
       "      <th>MoGene_Var_g</th>\n",
       "      <th>MoGene_SEg</th>\n",
       "      <th>RNA_Seq_Enrich</th>\n",
       "      <th>RNA_Seq_Hedges_g</th>\n",
       "      <th>...</th>\n",
       "      <th>W*5</th>\n",
       "      <th>esW*1</th>\n",
       "      <th>esW*2</th>\n",
       "      <th>esW*3</th>\n",
       "      <th>esW*4</th>\n",
       "      <th>esW*5</th>\n",
       "      <th>REM_M*</th>\n",
       "      <th>REM_SEM*</th>\n",
       "      <th>REM_Z</th>\n",
       "      <th>REM_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gm25122</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000219</td>\n",
       "      <td>0.087321</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.525325</td>\n",
       "      <td>10.965121</td>\n",
       "      <td>0.959891</td>\n",
       "      <td>3.371099e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm24428</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000219</td>\n",
       "      <td>0.087321</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.525325</td>\n",
       "      <td>10.965121</td>\n",
       "      <td>0.959891</td>\n",
       "      <td>3.371099e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hap1</th>\n",
       "      <td>1.876495</td>\n",
       "      <td>11.982360</td>\n",
       "      <td>2.764375</td>\n",
       "      <td>1.662641</td>\n",
       "      <td>2.555813</td>\n",
       "      <td>13.350462</td>\n",
       "      <td>3.869200</td>\n",
       "      <td>1.967028</td>\n",
       "      <td>2.769519</td>\n",
       "      <td>10.679511</td>\n",
       "      <td>...</td>\n",
       "      <td>0.051568</td>\n",
       "      <td>0.644864</td>\n",
       "      <td>0.678168</td>\n",
       "      <td>0.253540</td>\n",
       "      <td>0.394390</td>\n",
       "      <td>0.550719</td>\n",
       "      <td>9.359316</td>\n",
       "      <td>1.926535</td>\n",
       "      <td>4.858108</td>\n",
       "      <td>1.185128e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Baiap3</th>\n",
       "      <td>2.661205</td>\n",
       "      <td>13.108139</td>\n",
       "      <td>3.268774</td>\n",
       "      <td>1.807975</td>\n",
       "      <td>3.291131</td>\n",
       "      <td>9.606490</td>\n",
       "      <td>2.078571</td>\n",
       "      <td>1.441725</td>\n",
       "      <td>3.867353</td>\n",
       "      <td>6.108862</td>\n",
       "      <td>...</td>\n",
       "      <td>0.052065</td>\n",
       "      <td>0.619553</td>\n",
       "      <td>0.481113</td>\n",
       "      <td>0.167335</td>\n",
       "      <td>0.529667</td>\n",
       "      <td>0.318056</td>\n",
       "      <td>8.899008</td>\n",
       "      <td>2.050886</td>\n",
       "      <td>4.339105</td>\n",
       "      <td>1.430640e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calcr</th>\n",
       "      <td>3.194799</td>\n",
       "      <td>9.188330</td>\n",
       "      <td>1.708097</td>\n",
       "      <td>1.306942</td>\n",
       "      <td>3.229606</td>\n",
       "      <td>6.631518</td>\n",
       "      <td>1.072162</td>\n",
       "      <td>1.035453</td>\n",
       "      <td>3.961542</td>\n",
       "      <td>10.009732</td>\n",
       "      <td>...</td>\n",
       "      <td>0.143188</td>\n",
       "      <td>1.663221</td>\n",
       "      <td>1.356559</td>\n",
       "      <td>1.182258</td>\n",
       "      <td>0.985824</td>\n",
       "      <td>1.433269</td>\n",
       "      <td>8.613022</td>\n",
       "      <td>1.140543</td>\n",
       "      <td>7.551685</td>\n",
       "      <td>4.296621e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sytl4</th>\n",
       "      <td>3.683219</td>\n",
       "      <td>14.215653</td>\n",
       "      <td>3.809157</td>\n",
       "      <td>1.951706</td>\n",
       "      <td>2.743443</td>\n",
       "      <td>11.585671</td>\n",
       "      <td>2.952386</td>\n",
       "      <td>1.718251</td>\n",
       "      <td>3.589678</td>\n",
       "      <td>9.060526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041372</td>\n",
       "      <td>0.560883</td>\n",
       "      <td>0.473109</td>\n",
       "      <td>0.224602</td>\n",
       "      <td>0.113903</td>\n",
       "      <td>0.374849</td>\n",
       "      <td>8.346457</td>\n",
       "      <td>2.185555</td>\n",
       "      <td>3.818919</td>\n",
       "      <td>1.340379e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Celf6</th>\n",
       "      <td>1.872957</td>\n",
       "      <td>7.538747</td>\n",
       "      <td>1.215370</td>\n",
       "      <td>1.102438</td>\n",
       "      <td>2.016121</td>\n",
       "      <td>8.986032</td>\n",
       "      <td>1.838241</td>\n",
       "      <td>1.355817</td>\n",
       "      <td>2.956878</td>\n",
       "      <td>9.736804</td>\n",
       "      <td>...</td>\n",
       "      <td>0.313010</td>\n",
       "      <td>5.380424</td>\n",
       "      <td>4.439709</td>\n",
       "      <td>3.493510</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.047714</td>\n",
       "      <td>7.905053</td>\n",
       "      <td>0.695093</td>\n",
       "      <td>11.372660</td>\n",
       "      <td>5.721708e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ahi1</th>\n",
       "      <td>1.616307</td>\n",
       "      <td>8.792829</td>\n",
       "      <td>1.581105</td>\n",
       "      <td>1.257420</td>\n",
       "      <td>2.206488</td>\n",
       "      <td>11.359352</td>\n",
       "      <td>2.844201</td>\n",
       "      <td>1.686476</td>\n",
       "      <td>2.309227</td>\n",
       "      <td>8.359601</td>\n",
       "      <td>...</td>\n",
       "      <td>0.136392</td>\n",
       "      <td>1.324839</td>\n",
       "      <td>1.437893</td>\n",
       "      <td>0.954501</td>\n",
       "      <td>0.585792</td>\n",
       "      <td>1.140184</td>\n",
       "      <td>7.706141</td>\n",
       "      <td>1.189847</td>\n",
       "      <td>6.476583</td>\n",
       "      <td>9.382281e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Usp29</th>\n",
       "      <td>1.447292</td>\n",
       "      <td>7.372962</td>\n",
       "      <td>1.171225</td>\n",
       "      <td>1.082231</td>\n",
       "      <td>2.397659</td>\n",
       "      <td>8.083968</td>\n",
       "      <td>1.517444</td>\n",
       "      <td>1.231846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.295087</td>\n",
       "      <td>5.327359</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.682687</td>\n",
       "      <td>0.813032</td>\n",
       "      <td>9.449424</td>\n",
       "      <td>3.407008e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Magel2</th>\n",
       "      <td>2.113392</td>\n",
       "      <td>10.151008</td>\n",
       "      <td>2.040553</td>\n",
       "      <td>1.428479</td>\n",
       "      <td>1.622476</td>\n",
       "      <td>7.689045</td>\n",
       "      <td>1.387670</td>\n",
       "      <td>1.177994</td>\n",
       "      <td>0.006425</td>\n",
       "      <td>0.569601</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032091</td>\n",
       "      <td>0.307870</td>\n",
       "      <td>0.237912</td>\n",
       "      <td>0.258524</td>\n",
       "      <td>0.322189</td>\n",
       "      <td>0.018279</td>\n",
       "      <td>7.671147</td>\n",
       "      <td>2.588633</td>\n",
       "      <td>2.963398</td>\n",
       "      <td>3.042633e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ngb</th>\n",
       "      <td>2.421108</td>\n",
       "      <td>8.261588</td>\n",
       "      <td>1.419319</td>\n",
       "      <td>1.191352</td>\n",
       "      <td>2.914707</td>\n",
       "      <td>10.375626</td>\n",
       "      <td>2.398758</td>\n",
       "      <td>1.548792</td>\n",
       "      <td>3.404732</td>\n",
       "      <td>8.217836</td>\n",
       "      <td>...</td>\n",
       "      <td>0.162312</td>\n",
       "      <td>1.537493</td>\n",
       "      <td>1.633223</td>\n",
       "      <td>0.896828</td>\n",
       "      <td>0.799252</td>\n",
       "      <td>1.333852</td>\n",
       "      <td>7.667505</td>\n",
       "      <td>1.112010</td>\n",
       "      <td>6.895179</td>\n",
       "      <td>5.379683e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zcchc12</th>\n",
       "      <td>1.414539</td>\n",
       "      <td>7.720604</td>\n",
       "      <td>1.264924</td>\n",
       "      <td>1.124688</td>\n",
       "      <td>2.894750</td>\n",
       "      <td>11.943615</td>\n",
       "      <td>3.127848</td>\n",
       "      <td>1.768572</td>\n",
       "      <td>2.746637</td>\n",
       "      <td>6.684972</td>\n",
       "      <td>...</td>\n",
       "      <td>0.170422</td>\n",
       "      <td>1.379241</td>\n",
       "      <td>1.600882</td>\n",
       "      <td>0.928094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.139264</td>\n",
       "      <td>7.600361</td>\n",
       "      <td>1.227099</td>\n",
       "      <td>6.193762</td>\n",
       "      <td>5.874506e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm12245</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.032251</td>\n",
       "      <td>0.188735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.596199</td>\n",
       "      <td>5.862946</td>\n",
       "      <td>1.295628</td>\n",
       "      <td>1.951036e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Snord47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007617</td>\n",
       "      <td>0.139596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.557949</td>\n",
       "      <td>7.165210</td>\n",
       "      <td>1.054812</td>\n",
       "      <td>2.915113e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm25789</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007617</td>\n",
       "      <td>0.139596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.557949</td>\n",
       "      <td>7.165210</td>\n",
       "      <td>1.054812</td>\n",
       "      <td>2.915113e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mir5117</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007617</td>\n",
       "      <td>0.139596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.557949</td>\n",
       "      <td>7.165210</td>\n",
       "      <td>1.054812</td>\n",
       "      <td>2.915113e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm24489</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007617</td>\n",
       "      <td>0.139596</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.557949</td>\n",
       "      <td>7.165210</td>\n",
       "      <td>1.054812</td>\n",
       "      <td>2.915113e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vwa5b1</th>\n",
       "      <td>2.413478</td>\n",
       "      <td>6.653628</td>\n",
       "      <td>0.991050</td>\n",
       "      <td>0.995515</td>\n",
       "      <td>1.001080</td>\n",
       "      <td>3.945120</td>\n",
       "      <td>0.480224</td>\n",
       "      <td>0.692982</td>\n",
       "      <td>2.553127</td>\n",
       "      <td>13.505128</td>\n",
       "      <td>...</td>\n",
       "      <td>0.065480</td>\n",
       "      <td>0.623127</td>\n",
       "      <td>0.388033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.884321</td>\n",
       "      <td>7.361369</td>\n",
       "      <td>1.970695</td>\n",
       "      <td>3.735417</td>\n",
       "      <td>1.874044e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arhgap36</th>\n",
       "      <td>3.203549</td>\n",
       "      <td>7.041778</td>\n",
       "      <td>1.085976</td>\n",
       "      <td>1.042102</td>\n",
       "      <td>3.573682</td>\n",
       "      <td>6.929499</td>\n",
       "      <td>1.156348</td>\n",
       "      <td>1.075336</td>\n",
       "      <td>3.743120</td>\n",
       "      <td>7.675538</td>\n",
       "      <td>...</td>\n",
       "      <td>0.511932</td>\n",
       "      <td>6.484285</td>\n",
       "      <td>5.992570</td>\n",
       "      <td>3.450034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.929354</td>\n",
       "      <td>7.238951</td>\n",
       "      <td>0.603795</td>\n",
       "      <td>11.989094</td>\n",
       "      <td>4.053148e-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gpld1</th>\n",
       "      <td>0.997211</td>\n",
       "      <td>6.613826</td>\n",
       "      <td>0.981620</td>\n",
       "      <td>0.990767</td>\n",
       "      <td>1.432149</td>\n",
       "      <td>12.326395</td>\n",
       "      <td>3.321391</td>\n",
       "      <td>1.822468</td>\n",
       "      <td>1.726946</td>\n",
       "      <td>9.143721</td>\n",
       "      <td>...</td>\n",
       "      <td>0.079471</td>\n",
       "      <td>0.607599</td>\n",
       "      <td>0.932056</td>\n",
       "      <td>0.545201</td>\n",
       "      <td>0.254039</td>\n",
       "      <td>0.726659</td>\n",
       "      <td>7.162711</td>\n",
       "      <td>1.528566</td>\n",
       "      <td>4.685903</td>\n",
       "      <td>2.787279e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows Ã— 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          MoEx_Enrich  MoEx_Hedges_g  MoEx_Var_g  MoEx_SEg  MoGene_Enrich  \\\n",
       "Gm25122           NaN            NaN         NaN       NaN            NaN   \n",
       "Gm24428           NaN            NaN         NaN       NaN            NaN   \n",
       "Hap1         1.876495      11.982360    2.764375  1.662641       2.555813   \n",
       "Baiap3       2.661205      13.108139    3.268774  1.807975       3.291131   \n",
       "Calcr        3.194799       9.188330    1.708097  1.306942       3.229606   \n",
       "Sytl4        3.683219      14.215653    3.809157  1.951706       2.743443   \n",
       "Celf6        1.872957       7.538747    1.215370  1.102438       2.016121   \n",
       "Ahi1         1.616307       8.792829    1.581105  1.257420       2.206488   \n",
       "Usp29        1.447292       7.372962    1.171225  1.082231       2.397659   \n",
       "Magel2       2.113392      10.151008    2.040553  1.428479       1.622476   \n",
       "Ngb          2.421108       8.261588    1.419319  1.191352       2.914707   \n",
       "Zcchc12      1.414539       7.720604    1.264924  1.124688       2.894750   \n",
       "Gm12245           NaN            NaN         NaN       NaN            NaN   \n",
       "Snord47           NaN            NaN         NaN       NaN            NaN   \n",
       "Gm25789           NaN            NaN         NaN       NaN            NaN   \n",
       "Mir5117           NaN            NaN         NaN       NaN            NaN   \n",
       "Gm24489           NaN            NaN         NaN       NaN            NaN   \n",
       "Vwa5b1       2.413478       6.653628    0.991050  0.995515       1.001080   \n",
       "Arhgap36     3.203549       7.041778    1.085976  1.042102       3.573682   \n",
       "Gpld1        0.997211       6.613826    0.981620  0.990767       1.432149   \n",
       "\n",
       "          MoGene_Hedges_g  MoGene_Var_g  MoGene_SEg  RNA_Seq_Enrich  \\\n",
       "Gm25122               NaN           NaN         NaN             NaN   \n",
       "Gm24428               NaN           NaN         NaN             NaN   \n",
       "Hap1            13.350462      3.869200    1.967028        2.769519   \n",
       "Baiap3           9.606490      2.078571    1.441725        3.867353   \n",
       "Calcr            6.631518      1.072162    1.035453        3.961542   \n",
       "Sytl4           11.585671      2.952386    1.718251        3.589678   \n",
       "Celf6            8.986032      1.838241    1.355817        2.956878   \n",
       "Ahi1            11.359352      2.844201    1.686476        2.309227   \n",
       "Usp29            8.083968      1.517444    1.231846             NaN   \n",
       "Magel2           7.689045      1.387670    1.177994        0.006425   \n",
       "Ngb             10.375626      2.398758    1.548792        3.404732   \n",
       "Zcchc12         11.943615      3.127848    1.768572        2.746637   \n",
       "Gm12245               NaN           NaN         NaN             NaN   \n",
       "Snord47               NaN           NaN         NaN             NaN   \n",
       "Gm25789               NaN           NaN         NaN             NaN   \n",
       "Mir5117               NaN           NaN         NaN             NaN   \n",
       "Gm24489               NaN           NaN         NaN             NaN   \n",
       "Vwa5b1           3.945120      0.480224    0.692982        2.553127   \n",
       "Arhgap36         6.929499      1.156348    1.075336        3.743120   \n",
       "Gpld1           12.326395      3.321391    1.822468        1.726946   \n",
       "\n",
       "          RNA_Seq_Hedges_g      ...            W*5     esW*1     esW*2  \\\n",
       "Gm25122                NaN      ...            NaN       NaN       NaN   \n",
       "Gm24428                NaN      ...            NaN       NaN       NaN   \n",
       "Hap1             10.679511      ...       0.051568  0.644864  0.678168   \n",
       "Baiap3            6.108862      ...       0.052065  0.619553  0.481113   \n",
       "Calcr            10.009732      ...       0.143188  1.663221  1.356559   \n",
       "Sytl4             9.060526      ...       0.041372  0.560883  0.473109   \n",
       "Celf6             9.736804      ...       0.313010  5.380424  4.439709   \n",
       "Ahi1              8.359601      ...       0.136392  1.324839  1.437893   \n",
       "Usp29                  NaN      ...            NaN  6.295087  5.327359   \n",
       "Magel2            0.569601      ...       0.032091  0.307870  0.237912   \n",
       "Ngb               8.217836      ...       0.162312  1.537493  1.633223   \n",
       "Zcchc12           6.684972      ...       0.170422  1.379241  1.600882   \n",
       "Gm12245                NaN      ...            NaN       NaN       NaN   \n",
       "Snord47                NaN      ...            NaN       NaN       NaN   \n",
       "Gm25789                NaN      ...            NaN       NaN       NaN   \n",
       "Mir5117                NaN      ...            NaN       NaN       NaN   \n",
       "Gm24489                NaN      ...            NaN       NaN       NaN   \n",
       "Vwa5b1           13.505128      ...       0.065480  0.623127  0.388033   \n",
       "Arhgap36          7.675538      ...       0.511932  6.484285  5.992570   \n",
       "Gpld1             9.143721      ...       0.079471  0.607599  0.932056   \n",
       "\n",
       "             esW*3     esW*4     esW*5     REM_M*   REM_SEM*      REM_Z  \\\n",
       "Gm25122   0.000219  0.087321       NaN  10.525325  10.965121   0.959891   \n",
       "Gm24428   0.000219  0.087321       NaN  10.525325  10.965121   0.959891   \n",
       "Hap1      0.253540  0.394390  0.550719   9.359316   1.926535   4.858108   \n",
       "Baiap3    0.167335  0.529667  0.318056   8.899008   2.050886   4.339105   \n",
       "Calcr     1.182258  0.985824  1.433269   8.613022   1.140543   7.551685   \n",
       "Sytl4     0.224602  0.113903  0.374849   8.346457   2.185555   3.818919   \n",
       "Celf6     3.493510       NaN  3.047714   7.905053   0.695093  11.372660   \n",
       "Ahi1      0.954501  0.585792  1.140184   7.706141   1.189847   6.476583   \n",
       "Usp29          NaN       NaN       NaN   7.682687   0.813032   9.449424   \n",
       "Magel2    0.258524  0.322189  0.018279   7.671147   2.588633   2.963398   \n",
       "Ngb       0.896828  0.799252  1.333852   7.667505   1.112010   6.895179   \n",
       "Zcchc12   0.928094       NaN  1.139264   7.600361   1.227099   6.193762   \n",
       "Gm12245   0.032251  0.188735       NaN   7.596199   5.862946   1.295628   \n",
       "Snord47   0.007617  0.139596       NaN   7.557949   7.165210   1.054812   \n",
       "Gm25789   0.007617  0.139596       NaN   7.557949   7.165210   1.054812   \n",
       "Mir5117   0.007617  0.139596       NaN   7.557949   7.165210   1.054812   \n",
       "Gm24489   0.007617  0.139596       NaN   7.557949   7.165210   1.054812   \n",
       "Vwa5b1         NaN       NaN  0.884321   7.361369   1.970695   3.735417   \n",
       "Arhgap36  3.450034       NaN  3.929354   7.238951   0.603795  11.989094   \n",
       "Gpld1     0.545201  0.254039  0.726659   7.162711   1.528566   4.685903   \n",
       "\n",
       "              REM_pVal  \n",
       "Gm25122   3.371099e-01  \n",
       "Gm24428   3.371099e-01  \n",
       "Hap1      1.185128e-06  \n",
       "Baiap3    1.430640e-05  \n",
       "Calcr     4.296621e-14  \n",
       "Sytl4     1.340379e-04  \n",
       "Celf6     5.721708e-30  \n",
       "Ahi1      9.382281e-11  \n",
       "Usp29     3.407008e-21  \n",
       "Magel2    3.042633e-03  \n",
       "Ngb       5.379683e-12  \n",
       "Zcchc12   5.874506e-10  \n",
       "Gm12245   1.951036e-01  \n",
       "Snord47   2.915113e-01  \n",
       "Gm25789   2.915113e-01  \n",
       "Mir5117   2.915113e-01  \n",
       "Gm24489   2.915113e-01  \n",
       "Vwa5b1    1.874044e-04  \n",
       "Arhgap36  4.053148e-33  \n",
       "Gpld1     2.787279e-06  \n",
       "\n",
       "[20 rows x 59 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values('REM_M*', inplace=True, ascending=False)\n",
    "df.head(20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31684, 59)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl Gene ID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rpl22l1</th>\n",
       "      <td>ENSMUSG00000039221</td>\n",
       "      <td>ribosomal protein L22 like 1 [Source:MGI Symbo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm10597</th>\n",
       "      <td>ENSMUSG00000095611</td>\n",
       "      <td>predicted gene 10597 [Source:MGI Symbol;Acc:MG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ext1</th>\n",
       "      <td>ENSMUSG00000061731</td>\n",
       "      <td>exostoses (multiple) 1 [Source:MGI Symbol;Acc:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smcr7</th>\n",
       "      <td>ENSMUSG00000018599</td>\n",
       "      <td>Smith-Magenis syndrome chromosome region, cand...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm7792</th>\n",
       "      <td>ENSMUSG00000094722</td>\n",
       "      <td>predicted gene 7792 [Source:MGI Symbol;Acc:MGI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Ensembl Gene ID  \\\n",
       "MGI symbol                       \n",
       "Rpl22l1     ENSMUSG00000039221   \n",
       "Gm10597     ENSMUSG00000095611   \n",
       "Ext1        ENSMUSG00000061731   \n",
       "Smcr7       ENSMUSG00000018599   \n",
       "Gm7792      ENSMUSG00000094722   \n",
       "\n",
       "                                                  Description  \n",
       "MGI symbol                                                     \n",
       "Rpl22l1     ribosomal protein L22 like 1 [Source:MGI Symbo...  \n",
       "Gm10597     predicted gene 10597 [Source:MGI Symbol;Acc:MG...  \n",
       "Ext1        exostoses (multiple) 1 [Source:MGI Symbol;Acc:...  \n",
       "Smcr7       Smith-Magenis syndrome chromosome region, cand...  \n",
       "Gm7792      predicted gene 7792 [Source:MGI Symbol;Acc:MGI...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import ENSMBL mouse GIDs and descriptions again\n",
    "dfX=pd.read_table('../BioMart_Ensmbl_index/mart_export72_MGIsymbol.txt',index_col=[2])  \n",
    "dfX.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dfX.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_FINAL = df.join(dfX, how='left', sort=True)\n",
    "df_FINAL2 =df_FINAL[df_FINAL.index.duplicated()!=True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31684, 61)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_FINAL2.shape  # check same number of rows as input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_FINAL2.to_csv('input_files/meta_out_Symbol(REM).csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## The p-values from this study will be used to compute q-values using the qvals R package (or web frontend at http://qvalue.princeton.edu/ ) then returned:\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively qvalues as defined by https://github.com/nfusi/qvalue (3-term BSD licence HERE<https://github.com/nfusi/qvalue/blob/master/LICENSE.txt>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import scipy as sp\n",
    "from scipy import interpolate\n",
    "\n",
    "\n",
    "def estimate(pv, m=None, verbose=False, lowmem=False, pi0=None):\n",
    "    \"\"\"\n",
    "    Estimates q-values from p-values\n",
    "    Args\n",
    "    =====\n",
    "    m: number of tests. If not specified m = pv.size\n",
    "    verbose: print verbose messages? (default False)\n",
    "    lowmem: use memory-efficient in-place algorithm\n",
    "    pi0: if None, it's estimated as suggested in Storey and Tibshirani, 2003.\n",
    "         For most GWAS this is not necessary, since pi0 is extremely likely to be\n",
    "         1\n",
    "    \"\"\"\n",
    "    assert(pv.min() >= 0 and pv.max() <= 1), \"p-values should be between 0 and 1\"\n",
    "\n",
    "    original_shape = pv.shape\n",
    "    pv = pv.ravel()  # flattens the array in place, more efficient than flatten()\n",
    "\n",
    "    if m is None:\n",
    "        m = float(len(pv))\n",
    "    else:\n",
    "        # the user has supplied an m\n",
    "        m *= 1.0\n",
    "\n",
    "    # if the number of hypotheses is small, just set pi0 to 1\n",
    "    if len(pv) < 100 and pi0 is None:\n",
    "        pi0 = 1.0\n",
    "    elif pi0 is not None:\n",
    "        pi0 = pi0\n",
    "    else:\n",
    "        # evaluate pi0 for different lambdas\n",
    "        pi0 = []\n",
    "        lam = sp.arange(0, 0.90, 0.01)\n",
    "        counts = sp.array([(pv > i).sum() for i in sp.arange(0, 0.9, 0.01)])\n",
    "        for l in range(len(lam)):\n",
    "            pi0.append(counts[l]/(m*(1-lam[l])))\n",
    "\n",
    "        pi0 = sp.array(pi0)\n",
    "\n",
    "        # fit natural cubic spline\n",
    "        tck = interpolate.splrep(lam, pi0, k=3)\n",
    "        pi0 = interpolate.splev(lam[-1], tck)\n",
    "        if verbose:\n",
    "            print(\"qvalues pi0=%.3f, estimated proportion of null features \" % pi0)\n",
    "\n",
    "        if pi0 > 1:\n",
    "            if verbose:\n",
    "                print(\"got pi0 > 1 (%.3f) while estimating qvalues, setting it to 1\" % pi0)\n",
    "            pi0 = 1.0\n",
    "\n",
    "    assert(pi0 >= 0 and pi0 <= 1), \"pi0 is not between 0 and 1: %f\" % pi0\n",
    "\n",
    "    if lowmem:\n",
    "        # low memory version, only uses 1 pv and 1 qv matrices\n",
    "        qv = sp.zeros((len(pv),))\n",
    "        last_pv = pv.argmax()\n",
    "        qv[last_pv] = (pi0*pv[last_pv]*m)/float(m)\n",
    "        pv[last_pv] = -sp.inf\n",
    "        prev_qv = last_pv\n",
    "        for i in xrange(int(len(pv))-2, -1, -1):\n",
    "            cur_max = pv.argmax()\n",
    "            qv_i = (pi0*m*pv[cur_max]/float(i+1))\n",
    "            pv[cur_max] = -sp.inf\n",
    "            qv_i1 = prev_qv\n",
    "            qv[cur_max] = min(qv_i, qv_i1)\n",
    "            prev_qv = qv[cur_max]\n",
    "\n",
    "    else:\n",
    "        p_ordered = sp.argsort(pv)\n",
    "        pv = pv[p_ordered]\n",
    "        qv = pi0 * m/len(pv) * pv\n",
    "        qv[-1] = min(qv[-1], 1.0)\n",
    "\n",
    "        for i in xrange(len(pv)-2, -1, -1):\n",
    "            qv[i] = min(pi0*m*pv[i]/(i+1.0), qv[i+1])\n",
    "\n",
    "        # reorder qvalues\n",
    "        qv_temp = qv.copy()\n",
    "        qv = sp.zeros_like(qv)\n",
    "        qv[p_ordered] = qv_temp\n",
    "\n",
    "    # reshape qvalues\n",
    "    qv = qv.reshape(original_shape)\n",
    "\n",
    "    return qv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
